include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h

```diff
@@ -508,8 +509,76 @@ using __repacked_tuple_t = typename __repacked_tuple<T>::type;
 template <typename _ContainerOrIterable>
 using __value_t = typename __internal::__memobj_traits<_ContainerOrIterable>::value_type;
 
-//A contract for future class: <sycl::event or other event, a value or sycl::buffers...>
-//Impl details: inheretance (private) instead of aggregation for enabling the empty base optimization.
+template <typename _T>
+struct __accessor
+{
+    using __accessor_t = sycl::accessor<_T, 1, sycl::access::mode::read_write, __dpl_sycl::__target_device,
+                                        sycl::access::placeholder::false_t>;
+    __accessor_t m_acc;
+    _T* m_ptr;
+    bool m_usm;
+
+    auto
+    get_pointer() const //should be cached within a kernel
+    {
+        return m_usm ? m_ptr : &m_acc[0];
+    }
+};
+
+template <typename _ExecutionPolicy, typename _T>
+struct __storage
+{
+  private:
+    using __sycl_buffer_t = sycl::buffer<_T, 1>;
+    ::std::shared_ptr<__sycl_buffer_t> m_sycl_buf;
```

Comment:
Since we are exclusively using `m_sycl_buf` or `m_usm_buf`, it may be worth it to move these two variables into a union or `std::variant` so we only ever store one pointer in this class.

Reply:
This was tried before but rendered tricky when passing them to a kernel. Thus, the current solution uses smart pointers which should incur only minor overheads. The runtime performance is very close to the explicit mode proposed in #1106.

Reply:
Ah, I see. This makes sense, then.